aws/resource_aws_sns_topic.go

```diff
@@ -147,7 +147,7 @@ func resourceAwsSNSUpdateRefreshFunc(
 				// if the error contains the PrincipalNotFound message, we can retry
 				if strings.Contains(awsErr.Message(), "PrincipalNotFound") {
 					log.Printf("[DEBUG] Retrying AWS SNS Topic Update: %s", params)
-					return nil, "retrying", nil
+					return nil, "retrying", err
```

Comment:
Returning `err` here instead of `nil` will cause this WaitForState method to quit immediately. We have to return `nil` here because the _intent_ is to retry right here:

- https://github.com/terraform-providers/terraform-provider-aws/blob/068ff1beeb935dc6034b52a6ee38cef9f6fd60bd/vendor/github.com/hashicorp/terraform/helper/resource/state.go#L41

I agree with Radek, reworking this into `resource.Retry` may be better. I realize this isn't your code or change so let me know if you want to hand it off. Until `StateChangeConf` can surface "error before last", we're in this spot.